### PR TITLE
Retainer changes

### PR DESCRIPTION
**Pull request by muXxer**

### CODE REVIEW

**Review by alexsporn — APPROVED** [review]

File: error.go [hunk 1]
```diff
@@ -5,20 +5,43 @@ import (
...
 	// ErrUnknownAddrType gets returned for unknown address types.
 	ErrUnknownAddrType = ierrors.New("unknown address type")
-	// ErrUnknownInputType gets returned for unknown input types.
-	ErrUnknownInputType = ierrors.New("unknown input type")
 	// ErrUnknownOutputType gets returned for unknown output types.
 	ErrUnknownOutputType = ierrors.New("unknown output type")
 	// ErrUnknownTransactionEssenceType gets returned for unknown transaction essence types.
 	ErrUnknownTransactionEssenceType = ierrors.New("unknown transaction essence type")
+	// ErrUnknownTransactinType gets returned for unknown transaction types.
+	ErrUnknownTransactinType = ierrors.New("unknown transaction type")
```
Comment (muXxer) [inline error.go]: Typo

File: error.go [hunk 2]
```diff
@@ -5,20 +5,43 @@ import (
 )
 
 var (
+	ErrFailedToRetrieveInput = ierrors.New("failed to retrieve input references")
 	// ErrUnsupportedInputType gets returned for unsupported input types.
-	ErrUnsupportedInputType = ierrors.New("unsupported input type")
+	ErrUnknownInputType = ierrors.New("unsupported input type")
```
Comment (muXxer) [inline error.go]: Unknown in the name, unsupported in the message?

File: error.go [hunk 3]
```diff
@@ -5,20 +5,43 @@ import (
...
 	// ErrUnsupportedContextInputType gets returned for unsupported context input types.
-	ErrUnsupportedContextInputType = ierrors.New("unsupported context input type")
+	ErrUnknownContextInputType = ierrors.New("unsupported context input type")
```
Comment (muXxer) [inline error.go]: Unknown in the name, unsupported in the message?

File: error.go [hunk 4]
```diff
@@ -5,20 +5,43 @@ import (
...
 	// ErrUnknownTransactionEssenceType gets returned for unknown transaction essence types.
 	ErrUnknownTransactionEssenceType = ierrors.New("unknown transaction essence type")
+	// ErrUnknownTransactinType gets returned for unknown transaction types.
+	ErrUnknownTransactinType = ierrors.New("unknown transaction type")
 	// ErrUnknownUnlockType gets returned for unknown unlock.
 	ErrUnknownUnlockType = ierrors.New("unknown unlock type")
 	//ErrUnexpectedUnderlyingType gets returned for unknown input type of transaction.
 	ErrUnexpectedUnderlyingType = ierrors.New("unexpected underlying type provided by the interface")
+	// ErrCouldNotResolveBICInput gets returned when the BIC input could not be retrieved.
+	ErrCouldNotResolveBICInput = ierrors.New("could not retrieve BIC input")
```
Comment (muXxer) [inline error.go]: resolve or retrieve?

File: error.go [hunk 5]
```diff
@@ -5,20 +5,43 @@ import (
...
+	// ErrCouldNotResolveRewardInput gets returned when the reward input could not be retrieved.
+	ErrCouldNotResolveRewardInput = ierrors.New("could not retrieve reward input")
```
Comment (muXxer) [inline error.go]: resolve or retrieve?

File: error.go [hunk 6]
```diff
@@ -5,20 +5,43 @@ import (
...
+	// ErrCouldNorRetrieveCommitment gets returned when the commitment could not be retrieved.
+	ErrCouldNorRetrieveCommitment = ierrors.New("could not retrieve commitment")
```
Comment (muXxer) [inline error.go]: I like resolve more tbh.
Here is a typo as well.

File: error.go [hunk 7]
```diff
@@ -5,20 +5,43 @@ import (
...
+	// ErrCommitmentInputMissing gets returned when the commitment has not been provided when needed.
+	ErrCommitmentInputMissing = ierrors.New("commitment input required with reward or BIC input")
+	// ErrNoStakingFeature gets returned when the validator reward could not be claimed.
+	ErrNoStakingFeature = ierrors.New("validator reward claim failed due to no staking feature provided")
+	// ErrFailedToClaimValidatorReward gets returned when the validator reward could not be claimed.
+	ErrFailedToClaimValidatorReward = ierrors.New("validator staking claim failed")
```
Comment (muXxer) [inline error.go]: why validator rewards? shouldn't we simply call it "StakingReward"?

File: error.go [hunk 8]
```diff
@@ -5,20 +5,43 @@ import (
...
+	// ErrFailedToClaimDelegatorReward gets returned when the delegator reward could not be claimed.
+	ErrFailedToClaimDelegatorReward = ierrors.New("delegator staking claim failed")
```
Comment (muXxer) [inline error.go]: "DelegationReward"

File: error.go [hunk 9]
```diff
@@ -5,20 +5,43 @@ import (
...
+	// ErrUnlockBlockSignatureInvalid gets returned when the block signature unlock block signature is invalid.
+	ErrUnlockBlockSignatureInvalid = ierrors.New("block signature unlock block signature invalid")
```
Comment (muXxer) [inline error.go]: is mentioning block signature two times really correct?

File: error.go [hunk 10]
```diff
@@ -5,20 +5,43 @@ import (
...
+	// ErrInvalidNativeTokenSet gets returned when the provided native tokens are invalid.
+	ErrInvalidNativeTokenSet = ierrors.New("provided native tokens are invalid")
+	// ErrChainTransitionInvalid gets returned when the chain transition state failed.
+	ErrChainTransitionInvalid = ierrors.New("chain transition state failed")
```
Comment (muXxer) [inline error.go]: maybe remove the word state?

File: transaction.go [hunk 11]
```diff
@@ -33,6 +33,10 @@ var (
 	ErrInputOutputManaMismatch = ierrors.New("inputs and outputs do not contain the same amount of Mana")
 	// ErrInputCreationAfterTxCreation gets returned if an input has creation time after the transaction creation time.
 	ErrInputCreationAfterTxCreation = ierrors.New("input creation time after tx creation time")
+	// ErrUnknownTransactionEssenceType gets returned for unknown transaction essence types.
+	ErrUnknownTransactionEssenceType = ierrors.New("unknown transaction essence type")
+	//ErrUnexpectedUnderlyingTransactionType gets returned for unknown input type of transaction.
```
Comment (muXxer) [inline transaction.go]: the comment, the name and the errors message are all somehow different.
And the comment should have a whitespace at the beginning.

File: transaction.go [hunk 12]
```diff
@@ -98,7 +102,7 @@ func (t *Transaction) BICInputs() ([]*BlockIssuanceCreditInput, error) {
 		case *CommitmentInput, *RewardInput:
 			// ignore this type
 		default:
-			return nil, ErrUnexpectedUnderlyingType
+			return nil, ErrUnexpectedUnderlyingTransactionType
```
Comment (muXxer) [inline transaction.go]: Shouldn't that be something like ErrUnknownContextInput Type?

File: transaction.go [hunk 13]
```diff
@@ -114,7 +118,7 @@ func (t *Transaction) RewardInputs() ([]*RewardInput, error) {
 		case *CommitmentInput, *BlockIssuanceCreditInput:
 			// ignore this type
 		default:
-			return nil, ErrUnexpectedUnderlyingType
+			return nil, ErrUnexpectedUnderlyingTransactionType
```
Comment (muXxer) [inline transaction.go]: Shouldn't that be something like ErrUnknownContextInput Type?

File: error.go [hunk 14]
```diff
@@ -4,44 +4,44 @@ import (
...
-	// ErrInvalidNativeTokenSet gets returned when the provided native tokens are invalid.
-	ErrInvalidNativeTokenSet = ierrors.New("provided native tokens are invalid")
-	// ErrChainTransitionInvalid gets returned when the chain transition state failed.
-	ErrChainTransitionInvalid = ierrors.New("chain transition state failed")
-	// ErrInvalidManaAmount gets returned when the mana amount is invalid.
-	ErrInvalidManaAmount = ierrors.New("invalid mana amount, calculation error, or overflow")
+	ErrNoStakingFeature = ierrors.New("staking reward claim failed due to no staking feature provided")
+
+	// ErrFailedToClaimStakingReward gets returned when the validator reward could not be claimed.
+	ErrFailedToClaimStakingReward = ierrors.New("staking claim failed")
```
Comment (muXxer) [inline error.go]:
```suggestion
	ErrFailedToClaimStakingReward = ierrors.New("staking rewards claim failed")
```

File: error.go [hunk 15]
```diff
@@ -4,44 +4,44 @@ import (
...
+	// ErrFailedToClaimDelegationReward gets returned when the delegation reward could not be claimed.
+	ErrFailedToClaimDelegationReward = ierrors.New("delegation claim failed")
```
Comment (muXxer) [inline error.go]:
```suggestion
	ErrFailedToClaimDelegationReward = ierrors.New("delegation rewards claim failed")
```

**Pull request merged**